File: tree/src/main/java/team/dedica/logo/tree/plants/TreeParameters.java
package team.dedica.logo.tree.plants;

import processing.core.PVector;
import team.dedica.logo.tree.GrowthParameters;
import team.dedica.logo.tree.PlantSegment;

import static team.dedica.logo.tree.Util.random;

public class TreeParameters implements GrowthParameters<Tree> {

    public static final double MIN_DIAMETER_TO_GROW = 1;
    public static final double MIN_BRANCH_PROBABILITY = 0.15;

    /**
     * branch probability is multiplied with this factor each turn
     */
    public static final float BRANCH_PROBABILITY_FACTOR = 0.5f;

    /**
     * branch growth velocity reduction per cycle
     */
    public static final float VELOCITY_REDUCTION = 0.55f;
    public static final int MAX_STEM_DIAMETER = 20;

    public PlantSegment<Tree> getSeed(PVector origin, float scaleFactor) {
        return new PlantSegment<>(origin,
                new PVector(0, random(9, 12) * -1 * scaleFactor),
                ((random(15, MAX_STEM_DIAMETER)) * scaleFactor),
                scaleFactor,
                this
        );
    }



    public boolean canStillGrow(PlantSegment<Tree> plantSegment) {
        return plantSegment.diameter > MIN_DIAMETER_TO_GROW;
    }

    public float getVelocityReduction() {
        return VELOCITY_REDUCTION;
    }

    /**
     * slight bump into other direction
     * @return bump vector
     */
    public PVector bump() {
        float upwards = -0.8f;
        float downwards = 0.5f;
        PVector bump = new PVector(random(-1, 1), random(upwards, downwards));
        bump.mult(0.5f);
        return bump;
    }

    public float getRandomVelocityFactor(PlantSegment plantSegment) {
        return random(20, 70) * plantSegment.scaleFactor;
    }

    public boolean isWithinBounds(PVector location) {
        return location.x > 0 && location.y > 0;
    }

    public float getBranchProbabilityFactor() {
        return BRANCH_PROBABILITY_FACTOR;
    }

    public float getNewDiameterFactor() {
        return random(0.7f, 0.9f);
    }

    @Override
    public boolean canStillBranch(PlantSegment<Tree> plant) {
        return plant.branchProbability > MIN_BRANCH_PROBABILITY;
    }

    public PVector createGrowthVector(PlantSegment<Tree> plant) {
        PVector growthVector = new PVector(plant.velocity.x, plant.velocity.y);
        growthVector.normalize();
        growthVector.mult(getVelocityReduction()); //slower growth
        growthVector.add(bump());
        growthVector.mult(getRandomVelocityFactor(plant) * plant.scaleFactor);
        return growthVector;
    }

}

File: src/main/kotlin/team/dedica/logoart/items/tree/TreeParameters.kt
package team.dedica.logoart.items.tree

import processing.core.PVector
import team.dedica.logoart.Util
import team.dedica.logoart.items.GrowthParameters
import team.dedica.logoart.items.PlantSegment

internal class TreeParameters : GrowthParameters {

    override fun canStillGrow(plant: PlantSegment): Boolean {
        return plant.diameter > MIN_DIAMETER_TO_GROW
    }

    /**
     * slight bump into other direction
     *
     * @return bump vector
     */
    override fun bump(): PVector {
        val xDirection = 1.95f
        val yDirection = 0.95f
        val bump = PVector(Util.random(-xDirection, xDirection), Util.random(-yDirection, yDirection))
        bump.mult(0.6f)
        return bump
    }

    override fun getRandomVelocityFactor(plant: PlantSegment): Float {
        return Util.random(30f, 40f) * plant.scaleFactor
    }

    override fun isWithinBounds(location: PVector): Boolean {
        return location.x > 0 && location.y > 0
    }

    override fun getVelocityReduction(): Float {
        return velocityReduction
    }

    override fun getBranchProbabilityFactor(): Float {
        return branchProbabilityFactor
    }

    override fun getNewDiameterFactor(): Float = Util.random(0.7f, 0.9f)

    override fun canBranch(segment: PlantSegment): Boolean {
        //first segment cannot branch
        return if (segment.iteration < MIN_ITERATION_TO_BRANCH && segment.children.size > 1) false
        else segment.branchProbability > MIN_BRANCH_PROBABILITY
    }

    override fun createGrowthVector(plant: PlantSegment): PVector {
        val growthVector = PVector(plant.velocity.x, plant.velocity.y)
        growthVector.normalize()
        growthVector.add(bump())
        growthVector.mult(velocityReduction) //slower growth
        growthVector.mult(getRandomVelocityFactor(plant) * plant.scaleFactor)
        return growthVector
    }

    companion object {
        const val MIN_DIAMETER_TO_GROW = 0.5
        const val MIN_BRANCH_PROBABILITY = 0.25

        /**
         * branch probability is multiplied with this factor each turn
         * influences how many child branches a segment can have (the higher the more)
         */
        const val branchProbabilityFactor = 0.4f

        /**
         * branch growth velocity reduction per cycle
         */
        const val velocityReduction = 0.65f
        const val MAX_STEM_DIAMETER = 20
        const val MIN_ITERATION_TO_BRANCH = 6
    }
}
